### ISSUE TITLE
Yamagi Quake 2: Tremolo effect on sounds

### ISSUE THREAD

**Issue by Hiradur** [kcat/openal-soft#19]
**Setup:** Windows 7 x64, Yamagi Quake 2 v5.31
  
  
**Issue**:
I'm not sure what's going on but I would describe it as
- sounds alternating between full level and silence very quickly  
-- short recording of the issue in OGG Vorbis available here: http://ge.tt/7ml777O2/v/0 (will be deleted 14 days from now on)  
  
**Notes:**
This issue was introduced with OpenAL Soft v1.16.0. Older versions don't have any issues with this game.

Log with Loglevel=3:
```
AL lib: (II) alc_initconfig: Supported backends: mmdevapi, dsound, winmm, null, wave
AL lib: (II) ReadALConfig: Loading config C:\Users\Kersten\AppData\Roaming\alsoft.ini...
AL lib: (II) LoadConfigFromFile: found 'stereo-mode' = 'headphones'
AL lib: (II) LoadConfigFromFile: found 'hrtf_tables' = 'hiradur-%r.mhr'
AL lib: (II) GetConfigValue: Key disable-cpu-exts not found
AL lib: (II) FillCPUCaps: Detected max CPUID function: 0xa (ext. 0x80000008)
AL lib: (II) FillCPUCaps: Vendor ID: ""
AL lib: (II) FillCPUCaps: Name: "Intel(R) Core(TM)2 Quad CPU    Q6600  @ 2.40GHz"
AL lib: (II) FillCPUCaps: Extensions: +SSE +SSE2 -SSE4.1
AL lib: (II) GetConfigValue: Key rt-prio not found
AL lib: (II) GetConfigValue: Key resampler not found
AL lib: (II) GetConfigValue: Key trap-al-error not found
AL lib: (II) GetConfigValue: Key trap-alc-error not found
AL lib: (II) GetConfigValue: Key reverb/boost not found
AL lib: (II) GetConfigValue: Key reverb/emulate-eax not found
AL lib: (II) GetConfigValue: Key drivers not found
AL lib: (II) ALCmmdevProxy_messageHandler: Starting message thread
AL lib: (II) ALCmmdevProxy_messageHandler: Message thread initialization complete
AL lib: (II) ALCmmdevProxy_messageHandler: Starting message loop
AL lib: (II) alc_initconfig: Initialized backend "mmdevapi"
AL lib: (II) alc_initconfig: Added "mmdevapi" for playback
AL lib: (II) alc_initconfig: Initialized backend "dsound"
AL lib: (II) alc_initconfig: Added "dsound" for capture
AL lib: (II) GetConfigValue: Key excludefx not found
AL lib: (II) GetConfigValue: Key default-reverb not found
AL lib: (II) GetConfigValue: Key channels not found
AL lib: (II) GetConfigValue: Key sample-type not found
AL lib: (II) GetConfigValue: Key format not found
AL lib: (II) GetConfigValue: Key frequency not found
AL lib: (II) GetConfigValue: Key periods not found
AL lib: (II) GetConfigValue: Key period_size not found
AL lib: (II) GetConfigValue: Key sources not found
AL lib: (II) GetConfigValue: Key slots not found
AL lib: (II) GetConfigValue: Key sends not found
AL lib: (II) GetConfigValue: Key cf_level not found
AL lib: (II) GetConfigValue: Key midi/internal-synth not found
AL lib: (II) SSynth_create: Not using internal MIDI synth
AL lib: (II) ALCmmdevProxy_messageHandler: Got message 1024
AL lib: (II) alcOpenDevice: Created device 036E0260, "Lautsprecher (Realtek High Definition Audio)"
AL lib: (II) UpdateDeviceParams: Pre-reset: Stereo, Float, 44100hz, 1024 update size x4
AL lib: (II) GetConfigValue: Key hrtf not found
AL lib: (II) ALCmmdevProxy_messageHandler: Got message 1025
AL lib: (II) UpdateDeviceParams: Post-reset: Stereo, Float, 44100hz, 896 update size x4
AL lib: (II) GetConfigValue: Key layout_stereo not found
AL lib: (II) GetConfigValue: Key layout not found
AL lib: (II) UpdateDeviceParams: HRTF disabled
AL lib: (II) UpdateDeviceParams: BS2B disabled
AL lib: (II) GetConfigValue: Key wide-stereo not found
AL lib: (II) ALCmmdevProxy_messageHandler: Got message 1026
AL lib: (II) alcCreateContext: Created context 01FDFD60
AL lib: (II) ALCmmdevProxy_messageHandler: Got message 1029
AL lib: (II) add_device: Got device "Lautsprecher (Realtek High Definition Audio)", "{0.0.0.00000000}.{d6999160-f893-47c3-89bb-fcab18aea143}"
AL lib: (II) add_device: Got device "Lautsprecher (Realtek High Definition Audio)", "{0.0.0.00000000}.{d6999160-f893-47c3-89bb-fcab18aea143}"
AL lib: (II) FreeContext: 01FDFD60
AL lib: (II) ALCmmdevProxy_messageHandler: Got message 1027
AL lib: (II) FreeDevice: 036E0260
AL lib: (II) ALCmmdevProxy_messageHandler: Got message 1028
```

**Comment by kcat** [comment]
I'm not sure I hear the problem. It sounds like there's a machine gun firing which plays in quick short bursts (as expected), and other sounds play and pan as they should.

**Comment by Hiradur** [comment]
recording of the same scene with OpenAL Soft v1.15.1 where the issue does not appear: http://ge.tt/9UxFT7O2/v/0

The issue should become clear when comparing the two recordings (it's best to use headphones for this comparison).
However, I realized that my depiction of the issue was wrong. It is not a tremolo effect, instead some sounds very quickly pan between left and right speaker. I discovered this by merging the left and right channel of the stereo track into one mono track. In the mono track the issue disappears.

**Comment by kcat** [comment]
I wonder, is it due to the sounds just panning more? That is, where a sound would only pan a little bit between left and right in the working output, the change is much stronger in the problem one?

I can hear a difference between the two recordings, though I can't really notice anything outright wrong with either of them (without also seeing video to tell where the sounds should be). The only thing I can pick out is that it sounds like the panning in the first/"bad" recording is stronger, and if that's the case, I may know why.

Earlier versions of OpenAL Soft use the source's ```AL_REFERENCE_DISTANCE``` property as a kind of 'area size' for the sound. When the listener is within that distance, it blends between the purely directional sound and a centered/omnidirectional sound, so that when a sound slowly crosses through or near the listener, it slowly pans from one side to the other. However, this behavior was removed because it was inconsistent with other implementations, and HRTF rendering didn't do it either. Additionally, there's an ```AL_EXT_SOURCE_RADIUS``` extension that ostensibly does something like it (which OpenAL Soft doesn't currently support, but will eventuially).

The only thing is, I don't remember exactly when that behavior was removed. I had a quick look through the commits between 1.15.1 and 1.16 and didn't notice a commit removing it, but it's possible I just overlooked it. So I'm not 100% sure if that's the issue or not.

**Comment by Hiradur** [comment]
The recorded scene is the demo of the shareware version of Quake 2 which runs after starting the game. It's similar to demo1.dm from the release version but not the same. If you want I can try to upload a video.

``` AL_REFERENCE_DISTANCE ```
could explain this. The recordings were recorded with HRTF off but with HRTF on you can clearly hear how sounds originating from your own position (gunfire, screams, voice message) pan between left and right. With v1.15 they usually come from the center with a few exceptions (going out of bounds of that area?).
I find this a bit exhausting and tiresome after playing for a while, but as I see it, it should be fixed by the yquake developers, correct?

**Comment by kcat** [comment]
> The recordings were recorded with HRTF off but with HRTF on you can clearly hear how sounds originating from your own position (gunfire, screams, voice message) pan between left and right. With v1.15 they usually come from the center with a few exceptions (going out of bounds of that area?).

Ah, I see. So player/local sounds are panning when they should stay completely centered?

> I find this a bit exhausting and tiresome after playing for a while, but as I see it, it should be fixed by the yquake developers, correct?

I've had a look at their source, and yes, it seems to be an issue on their side. To play a local sound, they give it the same position as the listener, which works in theory but has two caveats:

* Asynchronous updates can cause the listener position to be updated before the source, or vice-versa, so there may be occasional periods rendered where they're out of sync. Unless batching/deferred updates are used, there's no way to fix this (and 1.16 has a potential bug with batching listener changes).
* Floating-point rounding errors with the math may cause the calculated local position to be slightly off from 0,0,0 even when the positions are the same. OpenAL Soft tries to account for this by testing that the distance is ```> FLT_EPSILON``` (about ```0.0000001192```) before applying panning, but it's possible this might not be enough in some cases (perhaps ```1.0f / 65536.0f``` would be better?).

Though the correct thing to do for local sounds is to set AL_SOURCE_RELATIVE to true on the source and set the position to 0,0,0, which will ensure the source doesn't audibly move from the listener.

**Comment by Hiradur** [comment]
> So player/local sounds are panning when they should stay completely centered?

Yes, that should be the issue. I wasn't sure at first since sounds pan so quickly between left and right channel that I was too confused to realize where the sounds originate from.


> Asynchronous updates can cause the listener position to be updated before the source, or vice-versa, so there may be occasional periods rendered where they're out of sync. Unless batching/deferred updates are used, there's no way to fix this (and 1.16 has a potential bug with batching listener changes).

This sounds about right, I sometimes had the feeling the sounds were lagging behind the camera's movement. Rounding errors may also play a role though.

Feel free to close this, I'll let the yquake developers know soon.

**Comment by Hiradur** [comment]
Closing as this indeed was an issue on yquake's side. It's now fixed.
